### PR TITLE
feat: add new channels object

### PR DESCRIPTION
**Pull request by fmvilas**
**Related issue(s):**
#663
#682  
#618

### CODE REVIEW

**Comment by fmvilas** [comment]
@GreenRover this PR should unblock your request/reply work.

File: spec/asyncapi.md
Comment (magicmatatjahu) [inline spec/asyncapi.md:652]: That change shouldn't be related to this PR, but shouldn't we also allow refs in channel's `servers`? I see that instead of IDs here references are used for channels and messages, so for servers they should too, like:

```yaml
servers:
  - $ref: '#/servers/rabbitmqInProd'
  - $ref: '#/servers/rabbitmqInStaging'
```

I think we even had a discussion on this topic once at some 3.0.0 meeting but I can't remember at which one.

WDYT? cc @smoya

Comment (smoya) [reply spec/asyncapi.md:652]: I think you mean my mention in https://github.com/asyncapi/spec/pull/806#discussion_r900012119 right?

Comment (magicmatatjahu) [reply spec/asyncapi.md:652]: Exactly, the same logic, but I remember that there was a discussion about it at some meeting about the servers themselves.

Comment (fmvilas) [reply spec/asyncapi.md:652]: Yeah, if we end up accepting this PR, we'll have to open an issue to change everything else too. But it also really depends on the output of https://github.com/asyncapi/spec/issues/663.

**Comment by fmvilas** [comment]
/progress 60

New `channels` shape is defined but it's still pending review by code owners.

**Comment by fmvilas** [comment]
@derberg @dalelane @smoya would you mind having a look?

File: spec/asyncapi.md
Comment (fmvilas) [inline spec/asyncapi.md:165]: Notice `channels` is not required anymore.

File: spec/asyncapi.md
Comment (fmvilas) [inline spec/asyncapi.md:580]: This is new. Instead of assuming the address delimiter is a `/` character, we make it custom. Should we define this in the bindings? I mean not making it customizable but strictly defined in the binding. For instance, the AMQP 0-9-1 binding will have a paragraph saying something like "Implementations MUST use the `.` (dot) character as a delimiter for the addresses.".

What do you think?

File: spec/asyncapi.md
Comment (fmvilas) [inline spec/asyncapi.md:579]: Address is no longer a URL template and, instead, it's just a plain string that supports Channel Address Expressions.

File: spec/asyncapi.md
Comment (fmvilas) [inline spec/asyncapi.md:581]: No more `oneOf`. The array has the "or" meaning by default.

File: spec/asyncapi.md
Comment (fmvilas) [inline spec/asyncapi.md:2588]: @derberg I fixed these two URLs and the Markdown links Action is still complaining. Any idea?

File: spec/asyncapi.md
Comment (fmvilas) [inline spec/asyncapi.md:670]: I went all-in defining our own templating rules here. I don't know what I was thinking when I defined the channel address as a URL template in version 2.0.0 😅 Have a look: https://www.rfc-editor.org/rfc/rfc6570#section-2.4.2.

File: spec/asyncapi.md
Comment (fmvilas) [inline spec/asyncapi.md:676]: This is covering most of the protocols supporting any form of wildcards: AMQP, MQTT, NATS, etc. It's still missing a Kafka feature that lets you subscribe to a regular expression. I tried to map it here but it's too complex and therefore left it out.

If you got any ideas on how to do it, I'm all ears. Otherwise, I'll open a follow-up issue and this can be included in future versions.

---

If you're curious, I came up with a solution like `users_{userId:/[\w\d]{8}/i}` but quickly got trapped in the [differences between regular expression engines](https://en.wikipedia.org/wiki/Comparison_of_regular_expression_engines) hole.

Kafka protocol doesn't define anything so it's hard for us to make a commitment here. Maybe go for the engine with fewer features, just to make sure everything is supported in every engine? Coming up with a subset ourselves? Should we just forget about it? Hot topic!

File: spec/asyncapi.md
Comment (fmvilas) [reply spec/asyncapi.md:580]: As a followup. I want this PR to be as much as possible about new channels object only.

File: spec/asyncapi.md
Comment (fmvilas) [reply spec/asyncapi.md:577]: Yeah, these helpers were needed because `/` is a delimiter in URLs but that's not the case anymore so whoever is designing the AsyncAPI file needs to make sure their addresses are valid for the protocol in question.

Agree it's not giving us much, gonna remove it. Thanks!

File: spec/asyncapi.md
Comment (fmvilas) [reply spec/asyncapi.md:2559]: LOL, funny. I tried it in [this commit](https://github.com/asyncapi/spec/pull/827/commits/82a335fd61fa376980c2460c5f5cb54021bca183) and didn't work. But the funniest thing is that there's a 404 on https://xml2rfc.ietf.org/public/rfc/html/rfc3339.html#anchor14 but that's exactly the URL I replaced in this PR. It shouldn't be checking for its existence, right?

**Comment by fmvilas** [comment]
@derberg I addressed all the feedback. Let me know what you think now.

File: spec/asyncapi.md
Comment (smoya) [reply spec/asyncapi.md:2559]: It seems the same issue is happening here as well: https://github.com/asyncapi/parser-js/pull/601#issuecomment-1240468666

File: spec/asyncapi.md
Comment (fmvilas) [reply spec/asyncapi.md:577]: Actually, thought it better and gonna address it here. It's still about the channels object.

File: spec/asyncapi.md
Comment (derberg) [reply spec/asyncapi.md:2562]: yeah, looks like there are some issues with action that we are using,
I'll open a PR to `.github` repo to use their new pre release

in the meantime, remove your changes of links as they will be fixed with https://github.com/asyncapi/spec/pull/832

**Review by derberg — APPROVED** [review]
LGTM

again, info, once we merge, we release a release candidate 😉

File: spec/asyncapi.md
Comment (jonaslagoni) [inline spec/asyncapi.md:568]: A shared communication channel with who? Am I the only one who feels like this sentence is incomplete in some way... 🤔

File: spec/asyncapi.md
Comment (jonaslagoni) [inline spec/asyncapi.md:577]:
```suggestion
<a name="channelObjectServers"></a>servers | [[Reference Object](#referenceObject)] | An array of `$ref` pointers to the definition of the servers in which this channel is available. If `servers` is absent or empty, this channel MUST be available on all the servers defined in the [Servers Object](#serversObject). Please note the `servers` property value MUST be an array of [Reference Objects](#referenceObject) and, therefore, MUST NOT contain an array of [Server Objects](#serverObject). However, it is RECOMMENDED that parsers (or other software) dereference this property for a better development experience.
```

**Review by jonaslagoni — COMMENTED** [review]
Just had a few remarks, otherwise looks good 👍

**Review by dalelane — APPROVED** [review]
LGTM - I particularly like moving the topic name out of the spec document key, and into an address field

File: spec/asyncapi.md
Comment (fmvilas) [reply spec/asyncapi.md:568]: We don't know in advance. We only know it's gonna be shared 🤷

File: spec/asyncapi.md
Comment (fmvilas) [reply spec/asyncapi.md]: Good catch 🦅 👁️

**Comment by fmvilas** [comment]
😱 😄

**Comment by fmvilas** [comment]
@derberg It didn't create any release candidate. Is it expected?

**Comment by derberg** [comment]
no, mainly because the fix I did in release pipeline did not go to `next-major-spec` branch as there are merge conflicts between it and `master`. It is good though, as PR title should be `feat!:` to get 3.0 RC and not 2.5 RC. So by accident we did not release a confusing release 😄

**Pull request merged**
